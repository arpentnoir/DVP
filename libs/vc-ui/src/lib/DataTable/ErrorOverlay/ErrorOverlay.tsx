/**
* @module ErrorOverlay
* @description This module exports the ErrorOverlay component which renders
  an error message overlay.
*/
import { Stack } from '@mui/material';
import { useEffect, useRef } from 'react';
import { DEFAULT_TABLE_ERROR_MESSAGE } from '../../constants';
import { Text } from '../../Text';

/**
* @typedef IErrorOverlay
* @type {object}
* @property {string} [errorMessage] - The error message to be displayed. Defaults to
  the value of DEFAULT_TABLE_ERROR_MESSAGE.
*/
interface IErrorOverlay {
  errorMessage?: string;
}

export const errorMessageTestId = 'text:table-error-message';

/**
 * @function ErrorOverlay
 * @description A functional React component that renders an error message overlay.
 * @param {IErrorOverlay} props - The props to configure the ErrorOverlay component.
 * @returns {JSX.Element} The rendered ErrorOverlay component.
 */
export const ErrorOverlay = ({
  errorMessage = DEFAULT_TABLE_ERROR_MESSAGE,
}: IErrorOverlay) => {
  const errorRef = useRef<HTMLElement>(null);

  useEffect(() => {
    if (errorMessage && errorRef?.current) {
      errorRef.current.focus();
    }
  }, [errorMessage]);

  return (
    <Stack justifyContent="center" alignItems="center" height="100%">
      <Text ref={errorRef} data-testid={errorMessageTestId}>
        {errorMessage}
      </Text>
    </Stack>
  );
};
