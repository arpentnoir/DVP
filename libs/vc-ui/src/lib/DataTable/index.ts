export * from './CursorPagination';
export * from './DataTable';
export * from './ErrorOverlay';
export * from './MoreInfo';
export * from './Toolbar';
