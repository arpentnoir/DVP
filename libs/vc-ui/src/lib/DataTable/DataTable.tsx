/**
* @module DataTable
* @description This module exports the DataTable component which renders
  a data table with a custom toolbar, pagination, and error handling.
*/
import { Box, BoxProps } from '@mui/material';
import { DataGrid, DataGridProps, GridSortModel } from '@mui/x-data-grid';
import { useCallback } from 'react';
import {
  CursorPagination,
  PaginationControls,
} from './CursorPagination/CursorPagination';
import { ErrorOverlay } from './ErrorOverlay';
import { Toolbar } from './Toolbar';

declare module '@mui/x-data-grid' {
  interface NoRowsOverlayPropsOverrides {
    errorMessage: string;
  }
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface PaginationPropsOverrides extends PaginationControls {}
}

type HandleSearch = (searchString: string) => Promise<void>;
type HandleSort = (sort: 'asc' | 'desc') => Promise<void>;

/**
 * @typedef IDataTable
 * @type {object}
 * @property {function} [handleAction] - The function to handle the toolbar button click action.
 * @property {string} [toolBarActionLabel] - The label for the toolbar action button.
 * @property {HandleSearch} [handleOnSearch] - The function to handle the toolbar search events.
 * @property {number} [minSearchQueryLength] - The minimum search query length before triggering a search.
 * @property {PaginationControls} [paginationControls] - The pagination controls for the data table.
 * @property {HandleSort} [handleOnSort] - The function to handle sorting events.
 * @property {boolean} [isLoading] - A flag indicating if the data is currently being fetched.
 * @property {string|null} [error] - An error message to be displayed, if any.
 * @property {BoxProps} [containerProps] - Props to be passed to the container Box component.
 */
interface IDataTable extends DataGridProps {
  handleAction?: () => void;
  toolBarActionLabel?: string;
  handleOnSearch?: HandleSearch;
  minSearchQueryLength?: number;
  paginationControls?: PaginationControls;
  handleOnSort?: HandleSort;
  isLoading?: boolean;
  error?: string | null;
  containerProps?: BoxProps;
}

/**
* @function DataTable
* @description A functional React component that renders a data table with
  a custom toolbar, pagination, and error handling.
* @param {IDataTable} props - The props to configure the DataTable component.
* @returns {JSX.Element} The rendered DataTable component.
*/
export const DataTable = ({
  handleAction,
  toolBarActionLabel,
  handleOnSearch,
  minSearchQueryLength,
  paginationControls,
  handleOnSort,
  isLoading,
  rows,
  error,
  containerProps,
  ...rest
}: IDataTable) => {
  const handleSort = useCallback(
    async (model: GridSortModel) => {
      if (handleOnSort && model?.length > 0) {
        const sort = model[0]?.sort ?? 'asc';
        await handleOnSort(sort);
      }
    },
    [handleOnSort]
  );

  return (
    <Box style={{ height: 600, width: '100%' }} {...containerProps}>
      <DataGrid
        slots={{
          // TODO: Swap loading overlay once built
          toolbar: Toolbar,
          pagination: CursorPagination,
          ...(error ? { noRowsOverlay: ErrorOverlay } : {}),
        }}
        slotProps={{
          toolbar: {
            handleAction,
            label: toolBarActionLabel,
            handleOnSearch,
            minSearchQueryLength,
          },
          pagination: paginationControls,
          ...(error ? { noRowsOverlay: { errorMessage: error } } : {}),
        }}
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        onSortModelChange={handleSort}
        sortingOrder={['asc', 'desc']}
        // TODO: Remove conditional and set to server when other pages are hooked up to the API
        sortingMode={handleOnSort ? 'server' : 'client'}
        rows={error ? [] : rows}
        loading={isLoading}
        {...rest}
      />
    </Box>
  );
};
