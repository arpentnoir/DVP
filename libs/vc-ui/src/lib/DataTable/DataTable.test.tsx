import { GridColDef } from '@mui/x-data-grid';
import { fireEvent, render, waitFor } from '@testing-library/react';
import {
  nextPageTestId,
  prevPageTestId,
} from './CursorPagination/CursorPagination';
import { DataTable } from './DataTable';
import { errorMessageTestId } from './ErrorOverlay';

const columns: GridColDef[] = [
  {
    field: 'id',
    headerName: 'ID',
  },
  {
    field: 'firstName',
    headerName: 'First Name',
  },
  {
    field: 'lastName',
    headerName: 'Last Name',
  },
];

const rows = [
  {
    id: '006',
    firstName: 'Postman',
    lastName: 'Pat',
  },
  {
    id: '007',
    firstName: 'Wreck-It',
    lastName: 'Ralph',
  },
];

const mockHandleAction = jest.fn();
const mockOnSearch = jest.fn();
const actionLabel = 'Issue';

const mockHandleNextPage = jest.fn();
const mockHandlePrevPage = jest.fn();

const mockHandleSort = jest.fn();

const mockPaginationControls = {
  handleNextPage: mockHandleNextPage,
  handlePrevPage: mockHandlePrevPage,
  nextCursor: null,
  prevCursor: null,
  limit: 50,
};

describe('DataTable', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    const { baseElement } = render(
      <DataTable
        columns={columns}
        rows={rows}
        handleOnSort={mockHandleSort}
        handleOnSearch={mockOnSearch}
        toolBarActionLabel={actionLabel}
        handleAction={mockHandleAction}
        paginationControls={mockPaginationControls}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should display all fields and their values', () => {
    const { getByText } = render(
      <DataTable
        columns={columns}
        rows={rows}
        handleOnSort={mockHandleSort}
        handleOnSearch={mockOnSearch}
        toolBarActionLabel={actionLabel}
        handleAction={mockHandleAction}
        paginationControls={mockPaginationControls}
      />
    );

    getByText('006');
    getByText('Postman');
    getByText('Pat');

    getByText('007');
    getByText('Wreck-It');
    getByText('Ralph');
  });

  describe('ToolbarComponent', () => {
    it('should use the handleOnSearch function on search', async () => {
      const { getByPlaceholderText } = render(
        <DataTable
          columns={columns}
          rows={rows}
          handleOnSort={mockHandleSort}
          handleOnSearch={mockOnSearch}
          toolBarActionLabel={actionLabel}
          handleAction={mockHandleAction}
          paginationControls={{
            ...mockPaginationControls,
            nextCursor: 'next',
            prevCursor: 'prev',
          }}
        />
      );

      expect(mockOnSearch).not.toHaveBeenCalled();

      const searchBar = getByPlaceholderText('Searchâ€¦');

      fireEvent.change(searchBar, { target: { value: '1234' } });

      await waitFor(() => {
        expect(mockOnSearch).toHaveBeenCalledTimes(1);
        expect(mockOnSearch).toHaveBeenCalledWith('1234');
      });
    });
  });

  describe('PaginationComponent', () => {
    it('should call the handleNextPage/handlePrevPage when their respective button is clicked', () => {
      const { getByTestId } = render(
        <DataTable
          columns={columns}
          rows={rows}
          handleOnSort={mockHandleSort}
          handleOnSearch={mockOnSearch}
          toolBarActionLabel={actionLabel}
          handleAction={mockHandleAction}
          paginationControls={{
            ...mockPaginationControls,
            nextCursor: 'next',
            prevCursor: 'prev',
          }}
        />
      );

      const prevPageIcon = getByTestId(prevPageTestId);
      const nextPageIcon = getByTestId(nextPageTestId);

      fireEvent.click(prevPageIcon);
      expect(mockHandlePrevPage).toBeCalledTimes(1);

      fireEvent.click(nextPageIcon);
      expect(mockHandleNextPage).toBeCalledTimes(1);
    });
  });

  describe('ErrorOverlayComponent', () => {
    it('should display error overlay when error is passed in', () => {
      const errorMessage = 'Test error message';

      const { queryByText, getByTestId } = render(
        <DataTable
          columns={columns}
          rows={rows}
          error={errorMessage}
          handleOnSort={mockHandleSort}
          handleOnSearch={mockOnSearch}
          toolBarActionLabel={actionLabel}
          handleAction={mockHandleAction}
          paginationControls={mockPaginationControls}
        />
      );

      expect(getByTestId(errorMessageTestId).textContent).toBe(errorMessage);
      expect(getByTestId(errorMessageTestId)).toHaveFocus();

      expect(queryByText('006')).toBeFalsy();
      expect(queryByText('Postman')).toBeFalsy();
      expect(queryByText('Pat')).toBeFalsy();

      expect(queryByText('007')).toBeFalsy();
      expect(queryByText('Wreck-It')).toBeFalsy();
      expect(queryByText('Ralph')).toBeFalsy();
    });
  });

  describe('SortingComponent', () => {
    it('should call the handleOnSort function provided when sorting', async () => {
      const { getByText } = render(
        <DataTable
          columns={columns}
          rows={rows}
          handleOnSort={mockHandleSort}
          handleOnSearch={mockOnSearch}
          toolBarActionLabel={actionLabel}
          handleAction={mockHandleAction}
          paginationControls={mockPaginationControls}
        />
      );

      const firstHeaderSlot = getByText('ID');

      // Order doesn't change on first click
      fireEvent.click(firstHeaderSlot);

      await waitFor(() => {
        expect(mockHandleSort).toBeCalledWith('asc');
        expect(mockHandleSort).toBeCalledTimes(1);
      });

      fireEvent.click(firstHeaderSlot);

      await waitFor(() => {
        expect(mockHandleSort).toBeCalledWith('desc');
        expect(mockHandleSort).toBeCalledTimes(2);
      });
    });
  });
});
