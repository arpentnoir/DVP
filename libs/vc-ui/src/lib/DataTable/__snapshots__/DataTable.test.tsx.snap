// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
      style="height: 600px; width: 100%;"
    >
      <div
        aria-colcount="3"
        aria-multiselectable="false"
        aria-rowcount="3"
        class="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-1usso8h-MuiDataGrid-root"
        role="grid"
      >
        <div>
          <div
            class="css-qz6wf1-MuiStack-root"
          >
            <div
              aria-label="Search"
              class="MuiFormControl-root MuiTextField-root css-3be3ve-MuiFormControl-root-MuiTextField-root-MuiDataGrid-toolbarQuickFilter"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1ptx2yq-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputTypeSearch MuiInputBase-inputAdornedEnd css-c63i49-MuiInputBase-input-MuiInput-input"
                  id=":r0:"
                  placeholder="Searchâ€¦"
                  type="search"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-ndac8y-MuiButtonBase-root-MuiButton-root"
              data-testid="button:Issue"
              style="margin-left: auto;"
              tabindex="0"
              type="button"
            >
              <div
                class="css-doi3ij-MuiStack-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  data-testid="button:Issue:label"
                  tabindex="0"
                >
                  Issue
                </p>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
        >
          <div
            class="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
            style="min-height: 56px; max-height: 56px; line-height: 56px;"
          >
            <div
              class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
              role="rowgroup"
              style="transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-rowindex="1"
                class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
                role="row"
              >
                <div
                  aria-colindex="1"
                  aria-label="ID"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor"
                  data-field="id"
                  role="columnheader"
                  style="height: 56px; width: 100px; min-width: 100px; max-width: 100px;"
                  tabindex="0"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          ID
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                      >
                        <button
                          aria-label="Sort"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                          tabindex="-1"
                          title="Sort"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                            data-testid="ArrowUpwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-menuIcon"
                    >
                      <button
                        aria-controls=":r3:"
                        aria-haspopup="true"
                        aria-label="Menu"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                        id=":r4:"
                        tabindex="-1"
                        title="Menu"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                          data-testid="TripleDotsVerticalIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                          />
                        </svg>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  aria-colindex="2"
                  aria-label="First Name"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor"
                  data-field="firstName"
                  role="columnheader"
                  style="height: 56px; width: 100px; min-width: 100px; max-width: 100px;"
                  tabindex="-1"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          First Name
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                      >
                        <button
                          aria-label="Sort"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                          tabindex="-1"
                          title="Sort"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                            data-testid="ArrowUpwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-menuIcon"
                    >
                      <button
                        aria-controls=":r6:"
                        aria-haspopup="true"
                        aria-label="Menu"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                        id=":r7:"
                        tabindex="-1"
                        title="Menu"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                          data-testid="TripleDotsVerticalIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                          />
                        </svg>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  aria-colindex="3"
                  aria-label="Last Name"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor"
                  data-field="lastName"
                  role="columnheader"
                  style="height: 56px; width: 100px; min-width: 100px; max-width: 100px;"
                  tabindex="-1"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          Last Name
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                      >
                        <button
                          aria-label="Sort"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                          tabindex="-1"
                          title="Sort"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                            data-testid="ArrowUpwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-menuIcon"
                    >
                      <button
                        aria-controls=":r9:"
                        aria-haspopup="true"
                        aria-label="Menu"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                        id=":ra:"
                        tabindex="-1"
                        title="Menu"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                          data-testid="TripleDotsVerticalIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                          />
                        </svg>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            style="flex: 1 1 0px; overflow: auto;"
          >
            <div
              class="MuiDataGrid-virtualScroller css-axafay-MuiDataGrid-virtualScroller"
              style="overflow-x: hidden;"
            >
              <div
                class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                style="width: auto; height: 104px; min-height: auto;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                  style="transform: translate3d(0px, 0px, 0px);"
                >
                  <div
                    aria-rowindex="2"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="006"
                    data-rowindex="0"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="id"
                      role="cell"
                      style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="006"
                      >
                        006
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="firstName"
                      role="cell"
                      style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Postman"
                      >
                        Postman
                      </div>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="lastName"
                      role="cell"
                      style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Pat"
                      >
                        Pat
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="3"
                    aria-selected="false"
                    class="MuiDataGrid-row MuiDataGrid-row--lastVisible"
                    data-id="007"
                    data-rowindex="1"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="id"
                      role="cell"
                      style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="007"
                      >
                        007
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="firstName"
                      role="cell"
                      style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Wreck-It"
                      >
                        Wreck-It
                      </div>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="lastName"
                      role="cell"
                      style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Ralph"
                      >
                        Ralph
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="Mui-resizeTriggers"
          >
            <div
              class="expand-trigger"
            >
              <div
                style="width: 1px; height: 1px;"
              />
            </div>
            <div
              class="contract-trigger"
            />
          </div>
        </div>
        <div
          class="MuiDataGrid-footerContainer MuiDataGrid-withBorderColor css-wop1k0-MuiDataGrid-footerContainer"
        >
          <div />
          <div
            class="MuiTablePagination-root css-1gl81ut-MuiTablePagination-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
              />
              <p
                class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
              >
                1â€“1 of 1
              </p>
              <div
                class="MuiTablePagination-actions"
              >
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="button:cursor-pagination-prev"
                  disabled=""
                  tabindex="-1"
                  title="Go to previous page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowLeftIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="button:cursor-pagination-next"
                  disabled=""
                  tabindex="-1"
                  title="Go to next page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowRightIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
